app/models/mixins/miq_provision_mixin.rb

```diff
@@ -139,34 +139,25 @@ def post_customization_templates(template_id)
 
   def set_folder(folder)
     return nil if folder.blank?
-    return set_resource(folder) if folder.kind_of?(MiqAeMethodService::MiqAeServiceEmsFolder)
 
-    result = nil
-    begin
-      if folder.kind_of?(Array) && folder.length == 2 && folder.first.kind_of?(Integer)
-        result = EmsFolder.find_by_id(folder.first)
-        result = [result.id, result.name] unless result.nil?
-      else
-        find_path = folder.to_miq_a.join('/')
-        result = get_folder_paths.detect { |_key, path| path.casecmp(find_path) == 0 }
-      end
-      update_attribute(:options, options.merge(:placement_folder_name => result)) unless result.nil?
-    rescue => err
-      _log.error "#{err}\n#{err.backtrace.join("\n")}"
+    result = if folder.kind_of?(MiqAeMethodService::MiqAeServiceEmsFolder)
+               folder
+             elsif folder.kind_of?(Array) && folder.length == 2 && folder.first.kind_of?(Integer)
+               MiqAeMethodService::MiqAeServiceEmsFolder.find(folder.first)
+             else
+               find_path = folder.to_miq_a.join('/')
+               eligible_resources(:folders).detect do |f|
+                 folder_path = f.folder_path(:exclude_root_folder => true, :exclude_non_display_folders => true)
+                 folder_path.casecmp(find_path).zero?
+               end
+             end
+
+    if result
+      result = MiqAeMethodService::MiqAeServiceEmsFolder.find(result.id)
```

Comment:
Will this only be needed for the last case?  It looks like [this](https://github.com/ManageIQ/manageiq/pull/13629/files#diff-0a8647613714b5b2dd896f1ceac5f7f0R143) and [this](https://github.com/ManageIQ/manageiq/pull/13629/files#diff-0a8647613714b5b2dd896f1ceac5f7f0R145) already return a `MiqAeMethodService::MiqAeServiceEmsFolder`.

Could you just do a find with the result of `eligible_resources(:folders).detect` and then just need `set_resource(result) if result` here and avoid an extra `find` in the other cases?